### PR TITLE
Add progress bar UI to download status tables to show download progress

### LINKED ISSUE #1308: Add progress indicators to downloads if the IDS supports it (ids.rd2foo)
**Description**:
Kevin's `ids.r2dfoo` exposes [getPercantageComplete](http://nexus.esc.rl.ac.uk:8081/nexus/content/sites/site/ids.r2dfoo/1.0.0/miredot/index.html#1826526913). This accepts `preparedId` and returns a string that is either parsable as an int `0-100`, or a string status indicating "UNKNOWN".

If the string is int parsable, we should render a progress bar with relevant percentage indicated, and if it's not parsable as an int we should just display the status message. We should only render a progress bar if the download is in the `RESTORING` state (double check this - we definitely don't need to render for `COMPLETE` and `EXPIRED` but are there other states?). We also need to test the case of a `preparedId` has been deleted from the topcat db and so `getPercantageComplete` returns 404.

question: Kevin says that a download can be 100% complete via `getPercantageComplete` but the download Availability is still not `COMPLETE` - this is due to pollcat delays etc. Should we do something to explain this to the user (add extra `Finalising` status to availability column?, add extra state to progress bar saying finalising, or do nothing and leave users to it)

This should also get added to the admin download table - we should investigate whether a bar is too wide for that table and whether to just display the percentage (ask Chris).

**Acceptance criteria**:
- [ ] Add extra column before the Availability column which displays progress bars

### PR DESCRIPTION
**Pull request by kennethnym**
## Description
This PR adds a new progress bar UI to both download status tables to show the progress of a download item. If the server returns a status string rather than a progress percentage, then the status string is shown instead.

I added a new entry in the settings file called `uiFeatures` which allow not only the progress UI to be toggled, but any future UI elements/features that need to be togglable:

```json
{
  "uiFeatures": {
    "downloadProgress": true
  }
}
```

It is an array of string flags that contains feature flags. The flags can then enable certain UI features, in this case download progress.

This should be merged after #1352 

## Testing instructions

- [ ] Review code
- [ ] Check Actions build
- [ ] Review changes to test coverage

## Agile board tracking
connect to #1308 

Closes #1308

### CODE REVIEW

File: packages/datagateway-download/public/datagateway-download-settings.example.json
Comment (louise-davies) [inline packages/datagateway-download/public/datagateway-download-settings.example.json:66]: A very minor point, but since this is JSON & Javascript normally the convention is `camelCase` and so `downloadProgress` may make more sense. Also, I dislike that this is just an array which needs a specific string - would an object which has boolean properties be clearer and less likely for someone to mess up by mispelling something?
i.e.

```
"uiFeatures": {
  "downloadProgress": true
}
```

If we did do things like that - I would then suggest that `downloadProgress` be disabled by default as this is a DLS only feature

File: packages/datagateway-download/src/downloadApi.ts
Comment (louise-davies) [inline packages/datagateway-download/src/downloadApi.ts:387]: I think `isNaN` is a bit unreliable - it has some odd results (`isNaN("")` is false, so is `isNaN([])`. I think the below code is more robust.

```
const numberData = parseInt(data); // or parseFloat - does the endpoint only return integers?
return Number.isNaN(numberData) ? data : numberData;
```

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (louise-davies) [inline packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:15]: This comment doesn't accurately portray that it takes a `downloadId`, not a `download`

File: packages/datagateway-download/public/datagateway-download-settings.example.json
Comment (kennethnym) [reply packages/datagateway-download/public/datagateway-download-settings.example.json:66]: Since it is a JSON file, I don't think the IDE would provide suggestion for the key anyways (unless we create a JSON schema). In TypeScript, I typed the array as an array of a union string, so in terms of autocompletion I think both are the same.

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (louise-davies) [inline packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:38]: Remove console.log statements

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (louise-davies) [inline packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: Why do we requery for the download here? Could we not just pass the required props from the status tables to save us requerying for all the downloads in the table?

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (louise-davies) [inline packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:58]: I thought that `LinearProgress` can be rendered with a label? https://mui.com/material-ui/react-progress/#linear-with-label

Comment (kennethnym) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:58]: If you open the code example, `LinearProgressWithLabel` is actually a custom component.

File: packages/datagateway-download/public/datagateway-download-settings.example.json
Comment (louise-davies) [reply packages/datagateway-download/public/datagateway-download-settings.example.json:66]: Yes, but from the perspective from a sysadmin who won't necessarily understand the source code - if they had a settings file with `"uiFeatures: []"` - how would they know what strings to put in? If this was `"uiFeatures": {"downloadProgress":false}` - they'd know to merely flip the boolean

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (louise-davies) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:58]: Ah okay - nevermind! I assumed it was a prop 😅

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (kennethnym) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: The problem is that the data in the table are formatted (`FormattedDownload`), which makes it awkward to work on the status of the download. In theory I can use a combination of `t` from useTranslation and translation key and use it to check the download status, but it feels wrong.

File: packages/datagateway-download/public/datagateway-download-settings.example.json
Comment (kennethnym) [reply packages/datagateway-download/public/datagateway-download-settings.example.json:66]: Ahh, I see what you mean now. I agree with you, I'll change it to an object type.

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (louise-davies) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: It feels like that would be more efficient than sending out more queries though right? Alternatively, we could do something like not use a select function to transform the downloads and instead do it in a separate variable (like what happens in `adminDownloadStatusTable` and then do something like this in the `cellContentRenderer`:

```
cellContentRenderer: ({
   rowData,
}: TableCellProps) => (
  <DownloadProgressIndicator
    download={data?.pages.flatMap((page) => page).find((download) => download.id === (rowData as FormattedDownload).id)}
  />
)
```

Also, couldn't you just make a hook that does the reverse of `formatDownload` - i.e. to transform from `FormattedDownload` to `Download` which you can use to abstract the `useTranslation` stuff and use the hook in `downloadProgressIndicator`?

Comment (kennethnym) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: Yeah, when I was writing that piece of code I was thinking calling `useDownload` there should have been a cache hit because `DownloadConfimDialog` also calls it, but then I just realized the table could've been loaded before the dialog (so obvious but I missed it).

What about a map that maps download IDs to `Download`, so that we can just look up the map for the unformatted `Download` when it is needed?

File: packages/datagateway-download/src/downloadApi.ts
Comment (kennethnym) [reply packages/datagateway-download/src/downloadApi.ts:387]: I can use `parseFloat` for now, it is easy to switch to `parseInt` if needed.

File: packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx
Comment (louise-davies) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: I don't see why the reverse of `formatDownload` is awkward, as we define both sides of the equation (theoretically there could be weird issues if the syadmin had duplicate string values in the translations but that's a niche concern I think). If you think having a map is more useful then go for it. A third alternative is to create a new field in `FormattedDownload` that's like `systemStatus` which keeps the original form of the status enum? For usages like this?

Comment (kennethnym) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: > theoretically there could be weird issues if the syadmin had duplicate string values in the translations but that's a niche concern I think
That is exactly what I am thinking about! What if the status values map to the same user-facing label? But if it is not a concern, then it will be way easier to just do a reverse of `formatDownload`.

Comment (kennethnym) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: I think I prefer adding an extra field to `FormattedDownload` - I was thinking adding a field called `formattedStatus` that stores the, well, formatted status, and the `status` will store the original value.

Comment (kennethnym) [reply packages/datagateway-download/src/downloadStatus/downloadProgressIndicator.component.tsx:22]: I can do the same for `isDeleted`.

**Review by louise-davies — COMMENTED** [review]
The code for this looks good - lets just have it tested next week & get feedback from the DLS admins

File: packages/datagateway-download/public/datagateway-download-settings.example.json
Comment (louise-davies) [inline packages/datagateway-download/public/datagateway-download-settings.example.json:66]: The example settings file & e2e settings haven't been updated to the new format

File: packages/datagateway-download/src/ConfigProvider.tsx
Comment (louise-davies) [inline packages/datagateway-download/src/ConfigProvider.tsx:30]: I don't know whether it's worth making this optional & just assuming false in that case for all flags - when I was deploying I forgot to update the settings file and it thus threw an exception...

Maybe this isn't a concern though as once it's in the new settings file it's considered a "standard" config value

File: packages/datagateway-download/src/downloadApiHooks.ts
Comment (louise-davies) [inline packages/datagateway-download/src/downloadApiHooks.ts:756]: Just had a thought - each download could have different IDSes and so we can't just get the "general" `idsUrl` from the settings file - we need the one corresponding to the download. We'll probably have to match the `transport` item in a download to the access methods array in the settings to get the right URL.

File: packages/datagateway-download/src/downloadApiHooks.ts
Comment (louise-davies) [inline packages/datagateway-download/src/downloadApiHooks.ts:767]: I'll need to check with Chris & Sabah whether they want progress request errors to be broadcast to the user or not (I've just asked Chris just waiting on a response)

**Comment by louise-davies** [comment]
@kennethnym I confirmed with Sabah that they're happy with this feature and they've been testing it pretty extensively. Once merge conflicts are resolved this can go in :)

File: packages/datagateway-download/public/res/default.json
Comment (louise-davies) [inline packages/datagateway-download/public/res/default.json:40]: A minor comment from yesterday was that `Unavailable` was confusing and they think perhaps `Unknown` is better. Might as well change the default before it merges in

**Review by louise-davies — APPROVED** [review]
Looks good and was well tested on DLS preprod :)

**Pull request merged**
